src/test/java/io/confluent/examples/streams/CustomJoinWithTableTriggeringStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,365 @@
+import org.apache.kafka.streams.kstream.ValueTransformerWithKey;
+import org.apache.kafka.streams.kstream.ValueTransformerWithKeySupplier;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.PunctuationType;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.Stores;
+import org.apache.kafka.test.TestUtils;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+
+import io.confluent.examples.streams.kafka.EmbeddedSingleNodeKafkaCluster;
+import io.confluent.examples.streams.utils.InstantSerde;
+import io.confluent.examples.streams.utils.Pair;
+import io.confluent.examples.streams.utils.PairOfDoubleAndLongDeserializer;
+import io.confluent.examples.streams.utils.PairSerde;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * End-to-end integration test that demonstrates one way to implement a custom join operation.
+ * Here, we implement custom Transformers with the Processor API and plug them into a DSL topology.
+ *
+ * Specifically, this example implements a stream-table join where both the stream side triggers a join output being
+ * sent downstream (default behavior of KStreams) but also the table side (not supported yet in Kafka Streams out of the
+ * box).  When a new record arrives on either side, a join output will be sent downstream immediately only if there is a
+ * matching record on the other side of the join.  If there is no such match, no join output will be sent immediately.
+ * Instead, the arriving record will be buffered to give the other side a certain (configurable) amount of time to,
+ * hopefully, see a matching record to arrive eventually.  However, if the per-key wait time has exceeded, then an
+ * "incomplete" join output will be sent downstream.
+ *
+ * Note: This example works with Java 8+ only.
+ */
+public class CustomJoinWithTableTriggeringStreamTableJoinIntegrationTest {
+
+  @ClassRule
+  public static final EmbeddedSingleNodeKafkaCluster CLUSTER = new EmbeddedSingleNodeKafkaCluster();
+
+  private static final String inputTopicForStream = "inputTopicForStream";
+  private static final String inputTopicForTable = "inputTopicForTable";
+  private static final String outputTopic = "outputTopic";
+  private static final String tableStoreName = "table-store";
+
+  @BeforeClass
+  public static void startKafkaCluster() {
+    CLUSTER.createTopic(inputTopicForStream);
+    CLUSTER.createTopic(inputTopicForTable);
+    CLUSTER.createTopic(outputTopic);
+  }
+
+  private static final class StreamTableJoinStreamSideTransformerSupplier
+    implements TransformerSupplier<String, Double, KeyValue<String, Pair<Double, Long>>> {
```

Comment:
I am using a `(String, Double) + (String, Long)  =>  (String, (Double, Long))` pipeline as example data types so that it is easier to adopt this example for your own use cases.  If the same data type is used multiple times (e.g. `String` for both key and value for one data source), the signatures of e.g. the Transformers are getting rather hard to read, and it's more difficult to understand which type is used for which part of the setup.